### PR TITLE
Add slicing, concatenation to StructuredGrid

### PR DESCRIPTION
**Pull request by whophil**
### Overview

Make working with `StructuredGrid` more natural.
- Add slicing via by implementing `StructuredGridFilters.extract_subset()`, calling vtkExtractGrid. Make this accessible from `StructuredGrid.__getitem__` (also retains legacy behavior)
    - Note that subsets are not views to the original data, which may be confusing if you expect everything to behave NumPy-like.
- Add concatenation by implementing `StructuredGridFilters.concatenate()`, calling vtkStructuredGridAppend

### Details

Demonstration of new features

```python
from pyvista import examples

structured = examples.load_structured()

# add some point data
structured.point_arrays.update({'my_field': structured.z.ravel(order='F')})

# this still works
structured['my_field']

# but you can also specify a numpy-like slice, "basic" indexing only
# here we split the grid into two
voi_1 = structured[:40, :40, :]
voi_2 = structured[40:, 40:, :]

# rejoin it
joined = voi_1.concatenate(voi_2)

print(structured)
print(joined)
```

Outputs:
```
StructuredGrid (0x14268e050)
  N Cells:	6241
  N Points:	6400
  X Bounds:	-1.000e+01, 9.750e+00
  Y Bounds:	-1.000e+01, 9.750e+00
  Z Bounds:	-1.000e+00, 1.000e+00
  Dimensions:	80, 80, 1
  N Arrays:	1

StructuredGrid (0x1426a1a60)
  N Cells:	6241
  N Points:	6400
  X Bounds:	-1.000e+01, 9.750e+00
  Y Bounds:	-1.000e+01, 9.750e+00
  Z Bounds:	-1.000e+00, 1.000e+00
  Dimensions:	80, 80, 1
  N Arrays:	1
```

### CODE REVIEW

**Comment by whophil** [comment]
Tests are failing on Linux non-conda. It does appear to be consistent across multiple builds, so I think it's real. but I'm unable to reproduce this in a Docker container based off of `docker/Dockerfile`. Any ideas?

**Comment by akaszynski** [comment]
> Tests are failing on Linux non-conda. It does appear to be consistent across multiple builds, so I think it's real. but I'm unable to reproduce this in a Docker container based off of `docker/Dockerfile`. Any ideas?

Give me a bit and I'll test it.  I'm using Python 3.7 on Linux.

**Comment by akaszynski** [comment]
Found the issue.  The original test splits the grid into a quarter:
```py
# split the grid into two
voi_1 = structured.extract_subset([0, 40, 0, 40, 0, 1], boundary=True)
```

To split the grid in half, it should read:
```py
voi_1 = structured.extract_subset([0, 80, 0, 40, 0, 1], boundary=True)
```

Patching and adding minor edits.

**Comment by whophil** [comment]
> Found the issue. The original test splits the grid into a quarter:
> 
> ```python
> # split the grid into two
> voi_1 = structured.extract_subset([0, 40, 0, 40, 0, 1], boundary=True)
> ```
> 
> To split the grid in half, it should read:
> 
> ```python
> voi_1 = structured.extract_subset([0, 80, 0, 40, 0, 1], boundary=True)
> ```
> 
> Patching and adding minor edits.

Thanks for finding this. Given that the inputs are bad, it is a little bit odd to me that the test passed on all but one of the build machines. As previously written, the two grids in the test were only joined at a corner. I don't think one should be able to concatenate these two grids at all, yet it seems VTK filter called allows it.

Should we do additional validation on the inputs in pyvista?

**Comment by whophil** [comment]
Due to the weird behavior I was seeing with `vtkStructuredGridAppend`, I rewrote `StructuredGrid.concatenate()` in Python. It is more explicit, requiring the concatenation axis to be specified. It also checks:
- grid dimensions are compatible for specified axis
- grids are coincident (within tolerance) along the seam
- all point arrays are equal along seam (no tolerance)

Due to the addition of the `axis` parameter, it's not clear what `struct_grid_1 + struct_grid_2` should do. So I took out the `StructuredGrid.__add__` method that you had implemented @akaszynski.

Added tests, but there could always be more.

**Comment by akaszynski** [comment]
The behavior with `vtkStructuredGridAppend` was quite strange, and there's clearly a bug in the VTK codebase.

I'm fine with removing the `__add__` method since you now need to add an axis for the concatenation.

**Comment by akaszynski** [comment]
When you get a chance, you may want to post an issue on VTK's gitlab site pointing out this behavior.

File: pyvista/core/filters.py [hunk 1]
```diff
@@ -4598,9 +4598,176 @@ def delaunay_2d(ugrid, tol=1e-05, alpha=0.0, offset=1.0, bound=False,
...
+            if dim1 != dim2:
+                raise RuntimeError('StructuredGrids with dimensions %s and %s '
+                                   'are not compatible.'
+                                   % (dataset.dimensions, other.dimensions))
+
+        # check point/cell variables are the same
+        assert set(dataset.point_arrays.keys()) == \
+               set(other.point_arrays.keys())
+        assert set(dataset.cell_arrays.keys()) == \
+               set(other.cell_arrays.keys())
```
Comment (akaszynski) [inline pyvista/core/filters.py:4716]: `assert` should only be used within CI.  There's a good SO answer here for why we should use `raise` instead:
https://stackoverflow.com/questions/20076182/what-are-the-implications-of-using-assert-vs-raise-exception

**Review by akaszynski — APPROVED** [review]
Approved.  I'd like to use `raise` rather than `assert`, but other than that I think we're good to go.  This wasn't an easy one; often times when working on `pyvista` I discover limitations with VTK and it's sometimes necessary to patch it within `pyvista`.

**Comment by whophil** [comment]
Thanks @akaszynski! As always, I appreciate the thorough review and response time in getting this through.

**Comment by akaszynski** [comment]
> Thanks @akaszynski! As always, I appreciate the thorough review and response time in getting this through.

You're welcome!  Now we've just got to this out as a release sometime soon.

**Pull request merged**